### PR TITLE
Rewrite the rest of the array non-generic fast paths in C#

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Array.cs [hunk 1]
```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
...
+                    case TypeCode.UInt32:
+                        if (TryGenericSort<uint>(keys, items, index, length)) return;
+                        break;
+                    case TypeCode.UInt64:
+                        if (TryGenericSort<ulong>(keys, items, index, length)) return;
+                        break;
+                }
 
-                    static bool TryGenericSort<TKey>(TKey[]? keys, Array? items, int index, int length)
+                static bool TryGenericSort<T>(Array keys, Array? items, int index, int length) where T : struct
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Array.cs]: The original FCalls was optimize only for the case where keys and items are the same type. I do not think we need to have a qudratic generic expansion here to do better.

File: src/System.Private.CoreLib/shared/System/Array.cs [hunk 2]
```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
...
-                            break;
-                        case TypeCode.UInt32:
-                            if (TryGenericSort(keys as uint[], items, index, length)) return;
-                            break;
-                        case TypeCode.UInt64:
-                            if (TryGenericSort(keys as ulong[], items, index, length)) return;
-                            break;
-                    }
+                    case TypeCode.Boolean:
+                        if (TryGenericSort<bool>(keys, items, index, length)) return;
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Array.cs]: I have moved the case inside the method to make it match the style I had to use for the other ones.

File: src/System.Private.CoreLib/shared/System/Array.cs [hunk 3]
```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
...
-                        case TypeCode.Int64:
-                            if (TryGenericSort(keys as long[], items, index, length)) return;
-                            break;
-                        case TypeCode.SByte:
-                            if (TryGenericSort(keys as sbyte[], items, index, length)) return;
-                            break;
-                        case TypeCode.Single:
-                            if (TryGenericSort(keys as float[], items, index, length)) return;
-                            break;
-                        case TypeCode.String:
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Array.cs]: Removed String. It was not optimized in the original FCall either and it does not have that bad performance either.

File: src/System.Private.CoreLib/shared/System/Array.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Array.cs]: Ok.

File: src/System.Private.CoreLib/shared/System/Array.cs [hunk 4]
```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
...
+            if (keys is object[] objKeys)
             {
-                comparer ??= Comparer.Default;
+                object[]? objItems = items as object[];
+                if (items == null || objItems != null)
+                {
+                    new SorterObjectArray(objKeys, objItems, comparer).Sort(index, length);
+                    return;
+                }
+            }
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Array.cs]: I'm curious why you moved this up front?

File: src/System.Private.CoreLib/shared/System/Array.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Array.cs]: I hadn't noticed that the FCall required them to be of the same type.  Happy for the simplification.

File: src/System.Private.CoreLib/shared/System/Array.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Array.cs]: `object[]` is fast check that does not involve reflection. I though that it makes sense to put it first before anything that involves reflection. But I may need to rethink how these switches are done. The original implementation had special casing for IntPtr/UIntPtr that I have lost with this change...

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

File: src/System.Private.CoreLib/shared/System/Array.cs [hunk 5]
```diff
@@ -949,23 +996,84 @@ public static int IndexOf(Array array, object? value, int startIndex, int count)
...
                     }
                 }
+
+                result = default;
+                return false;
+            }
+
+            for (int i = startIndex; i < endIndex; i++)
+            {
+                object? obj = array.GetValue(i);
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Array.cs]: `Array.GetValue` is relatively expensive API. The bounds checks are the least of your concern.

File: src/System.Private.CoreLib/shared/System/RuntimeType.cs [hunk 6]
```diff
@@ -187,8 +187,6 @@ protected override TypeCode GetTypeCodeImpl()
                     typeCode = TypeCode.Single; break;
                 case CorElementType.ELEMENT_TYPE_R8:
                     typeCode = TypeCode.Double; break;
-                case CorElementType.ELEMENT_TYPE_STRING:
-                    typeCode = TypeCode.String; break;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/RuntimeType.cs]: What drove this change?  This means that GetTypeCode will no longer return String for strings?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/RuntimeType.cs]: `RuntimeTypeHandle.GetCorElementType` does not ever return `ELEMENT_TYPE_STRING`. `ELEMENT_TYPE_STRING` is only used as a shortcut in signatures. It is not used anywhere else in the system.

`TypeCode.String` is returned via the default case.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/RuntimeType.cs]: Got it.

File: src/System.Private.CoreLib/shared/System/Array.cs [hunk 7]
```diff
@@ -1132,23 +1232,84 @@ public static int LastIndexOf(Array array, object? value, int startIndex, int co
...
+                result = default;
+                return false;
+            }
+
+            for (int i = startIndex; i >= endIndex; i--)
+            {
+                object? obj = array.GetValue(i);
+                if (obj == null)
+                {
+                    if (value == null)
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/shared/System/Array.cs]: I see a lot of `is null` in new code and here `== null` was added. Does the jit care which in this case? Is there a type or performance consideration to push in either direction?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Array.cs]: It matters only in the few rare cases where `==` operator is overload. There is no difference (the IL is exactly same) in 99% of cases.

**Pull request merged**